src/common/eigf.h

```diff
@@ -0,0 +1,302 @@
+ * just like what is done in fast_guided_filter.h
+**/
+
+static inline void exposure_independent_guided_filter(const float *const restrict guide, // I
+                                    const float *const restrict mask, //p
+                                    float *const restrict ab,
+                                    const size_t width, const size_t height,
+                                    const float sigma, const float feathering)
+{
+  // We also use gaussian blurs instead of the square blurs of the guided filter
+  const size_t Ndim = width * height;
+  float *const restrict blurred_guide = dt_alloc_sse_ps(Ndim);
+  // guide_x_guide = (guide - blurred_guide)^2
+  float *const restrict guide_x_guide = dt_alloc_sse_ps(Ndim);
+  // guide_variance = blur(guide_x_guide)
+  float *const restrict guide_variance = dt_alloc_sse_ps(Ndim);
+  float *const restrict guide_x_mask = dt_alloc_sse_ps(Ndim);
+  // guide_mask_covariance = blur(guide_x_mask)
+  float *const restrict guide_mask_covariance = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_mask = dt_alloc_sse_ps(Ndim);
+  float *const restrict a = dt_alloc_sse_ps(Ndim);
+  float *const restrict b = dt_alloc_sse_ps(Ndim);
+  // weight to compute the weighted blur of a and b
+  float *const restrict weights = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_a = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_b = dt_alloc_sse_ps(Ndim);
+  float *const restrict blurred_weights = dt_alloc_sse_ps(Ndim);
+
+  float ming = 10000000.0f;
+  float maxg = 0.0f;
+  float minm = 10000000.0f;
+  float maxm = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+dt_omp_firstprivate(guide, mask, Ndim) \
+  schedule(simd:static) aligned(guide, mask:64) \
+  reduction(max:maxg, maxm)\
+  reduction(min:ming, minm)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float pixelg = guide[k];
+    const float pixelm = mask[k];
+    if(pixelg < ming) ming = pixelg;
+    if(pixelg > maxg) maxg = pixelg;
+    if(pixelm < minm) minm = pixelm;
+    if(pixelm > maxm) maxm = pixelm;
+  }
+
+  dt_gaussian_t *g = dt_gaussian_init(width, height, 1, &maxg, &ming, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, guide, blurred_guide);
+  dt_gaussian_free(g);
+
+  g = dt_gaussian_init(width, height, 1, &maxm, &minm, sigma, 0);
+  if(!g) return;
+  dt_gaussian_blur(g, mask, blurred_mask);
+  dt_gaussian_free(g);
+
+  float mingg = 10000000.0f;
+  float maxgg = 0.0f;
+  float mingm = 10000000.0f;
+  float maxgm = 0.0f;
+#ifdef _OPENMP
+#pragma omp parallel for simd default(none) \
+  dt_omp_firstprivate(guide, mask, blurred_guide, blurred_mask, guide_x_guide, guide_x_mask, Ndim) \
+  schedule(simd:static) aligned(guide, mask, blurred_guide, blurred_mask, guide_x_guide, guide_x_mask:64) \
+  reduction(max:maxgg, maxgm)\
+  reduction(min:mingg, mingm)
+#endif
+  for(size_t k = 0; k < Ndim; k++)
+  {
+    const float deviation = guide[k] - blurred_guide[k];
```

Comment:
here, if you had one array of vectors `{ guide, mask, guide_x_guide, guide_x_mask }`, you could blur them horizontally all in that loop, with SIMD, then compute the vertical blur in another loop, where the covariance and (a, b) computation can happen after the convolution. That would save a lot of I∕O, take advantage of data alignment and avoid using the `dt_gaussian_xx` lib which is quite slow IIRC.

Reply:
Ok, I will use box averages instead of gaussian blurs and do that